proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,190 @@
+However, we have not received much feedback that this is desirable for package sources, so it's unlikely that it's important for audit sources.
+
+## Prior Art
+
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+Although [`npm audit`'s help docs](https://docs.npmjs.com/cli/v10/commands/npm-audit?v=true) don't mention it, it appears to have a `--registry={url}` argument, as can be seen on [the Azure DevOps docs on npm audit](https://learn.microsoft.com/azure/devops/artifacts/npm/npm-audit?view=azure-devops&tabs=yaml).
+
+Pypi's `pip` has [an audit command](https://pypi.org/project/pip-audit/), and `--index-url` and `--extra-index-url` appear to enable using custom audit sources.
+
+Rust's cargo audit command has [configuration for the advisory DB URL](https://docs.rs/cargo-audit/latest/cargo_audit/config/struct.DatabaseConfig.html)
+
+## Unresolved Questions
+
+<!-- What parts of the proposal do you expect to resolve before this gets accepted? -->
+<!-- What parts of the proposal need to be resolved before the proposal is stabilized? -->
+<!-- What related issues would you consider out of scope for this proposal but can be addressed in the future? -->
+
+1. default value?
+
+   Should we add nuget.org as a default value, so that customers who remove nuget.org as a package source are more easily able to use NuGetAudit?
+   Would using nuget.org as a default audit source cause undue concern from customers who don't want to use nuget.org as a package source, and the new tooling version starts making network requests to nuget.org?
+
+   If we want to make nuget.org a default/automatic audit source, how?
+   NuGet has tried before to use a tracking file to automatically add nuget's v3 URL as a package source to user-profile nuget.config files that don't have any package sources.
+   However, this caused a security issue when customers wanted to remove nuget.org as a default package source, and this code incorrectly re-added it once (the customer could successfully remove nuget.org a second time).
+   When we removed this auto-update to mitigate the security concern, it exposed a bug where multiple other apps created user-profile nuget.config files without any package sources, and many customers assumed it was a bug with NuGet.
+
+   Should we hard-code nuget.org's default URL in NuGet.Configuration, and require customers to use `<clear/>` if they don't want any audit source?
+
+   Should we just add it as a default in new user-profile nuget.config files, and just accept that existing developers won't get it by default?
+
+2. Stop using package source as audit source?
```

Comment:
Haven't you already answered this implicitly? The inverse of what you say in the quote below, I believe, is that *not* enabling any `auditSource`, means the current behavior remains. That is, continue looking for `VulnerabilityInfo` in the `packageSource` and if it's there, *do use it* from the packageSource's resource. In practical terms, that means `nuget.org` will be the VDB. No need to stop this from what I can tell.

> When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.

Reply:
@jeffkl wrote something related in his PR review comment, which GitHub infuriatingly doesn't make a "conversation", making it much harder to respond to and discuss (have a conversation about).

> Haven't you already answered this implicitly?

No. I wish I knew what it was about my writing that I could improve to reduce misunderstandings. Having said that, when I feel like I'm under time pressure, I don't read other peoples specs carefully enough to understand subtle differences either 😕 

What I intended in this section was specifically to make it so `packageSources` are **NEVER** used as audit sources. The earlier proposal to use packageSources when no auditSources are defined is solely for backwards compatibility reasons. If I could go back in time a few months to before NuGetAudit V1 was merged, I would implement it requiring auditSources as the only way to define which source provides audit info, no fallback to packageSources.

Backwards compatibility is important, but as NuGetAudit is a brand new feature, there's next to zero adoption. In fact, NuGetAudit isn't even GA yet, but it's too late to make any changes to what will ship as GA now. So whatever this spec proposes/implements will be changes to GA behaviour. But, being a new feature, the earlier we make "breaking" changes, the fewer customers will be impacted. If we wait until we customers start using the feature, and several customers give us feedback saying they think some other behaviour would be better, then by definition there will be a lot more people impacted by the change.  So, I thought there'd be opportunity to make a potential breaking change while adoption is low, so most people would never notice.

Reply:
Watched the meeting recording and it seems like decision were made. I'll re-review the latest iteration.